### PR TITLE
Speed up all the array.fs for loops

### PR DESCRIPTION
**Pull request by jackmott**
This addresses issue #1419 , where a pattern was used in for loops that leads to bad performance. This PR should speed up many functions in array.fs.

`array.rev` was also slightly refactored at the same time to make it much faster.

### CODE REVIEW

**Comment by jackmott** [comment]
ci error looks related to AsyncChoice, which I don't think I touched.
Related to issue #1379 ?

**Comment by jackmott** [comment]
I'm merging the latest master into this and cleaning some things up. Please don't merge yet.

File: src/fsharp/FSharp.Core/array.fs [hunk 1]
```diff
@@ -1136,23 +1128,22 @@ namespace Microsoft.FSharp.Collections
...
+                let inputLength = array.Length                
 
                 let isTrue = Array.zeroCreateUnchecked inputLength
                 Parallel.For(0, inputLength, 
                     fun i -> isTrue.[i] <- predicate array.[i]
                     ) |> ignore
                 
                 let mutable trueLength = 0
-                for i in 0 .. lastInputIndex do
+                for i in 0 .. isTrue.Length-1 do
```
Comment (asik) [inline src/fsharp/FSharp.Core/array.fs]: This is gets optimized in the same way, but you might as well use the `for i in 0 to array.Length-1` form everywhere for consistency.

Comment (jackmott) [inline src/fsharp/FSharp.Core/array.fs]: I was going by a pattern of referring to the first array to appear in the loop.  For instance
`indexed` uses res.Length-1 because res appears before array.  There is no logical reason for that, that was just my internal consistency .

Happy to change if if you like the array.Length-1 pattern better though

Comment (asik) [inline src/fsharp/FSharp.Core/array.fs]: I'm referring to the use of `for i in 0 .. bla` instead of `for i = 0 to bla`. There's no perf difference but you've used the latter consistently throughout the rest of the PR, so just thought it would make more sense.

Comment (jackmott) [inline src/fsharp/FSharp.Core/array.fs]: Oh! thanks. I'll adjust that to be consistent.  Is there any general community consensus on which is preferred?

Comment (asik) [inline src/fsharp/FSharp.Core/array.fs]: I couldn't say. `0 .. bla` it's a range expression so you can write fancier things, `0 to bla` is the low-level for loop that doesn't rely on too much compiler optimizations to be good. I prefer writing the latter when I want to make it clear I'm writing for performance.

Comment (jackmott) [inline src/fsharp/FSharp.Core/array.fs]: Probably saves the compiler some work too.

File: src/fsharp/FSharp.Core/array.fs [hunk 2]
```diff
@@ -748,7 +741,7 @@ namespace Microsoft.FSharp.Collections
             let mutable state = acc 
             let len = array1.Length
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/array.fs]: Remove ``len`` altogether here?

File: src/fsharp/FSharp.Core/array.fs [hunk 3]
```diff
@@ -283,7 +279,7 @@ namespace Microsoft.FSharp.Collections
             let f = OptimizedClosures.FSharpFunc<_,_,_>.Adapt(f)
             let len1 = array1.Length
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/array.fs]: remove ``len1`` altogether here?

File: src/fsharp/FSharp.Core/array.fs [hunk 4]
```diff
@@ -307,7 +303,7 @@ namespace Microsoft.FSharp.Collections
...
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/array.fs]: remove ``len1`` altogether here?

File: src/fsharp/FSharp.Core/array.fs [hunk 5]
```diff
@@ -333,16 +329,15 @@ namespace Microsoft.FSharp.Collections
...
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/array.fs]: remove ``len1`` altogether here?

File: src/fsharp/FSharp.Core/array.fs [hunk 6]
```diff
@@ -352,16 +347,15 @@ namespace Microsoft.FSharp.Collections
             let f = OptimizedClosures.FSharpFunc<_,_,_,_>.Adapt(f)
             let len1 = array1.Length
```
Comment (dsyme) [inline src/fsharp/FSharp.Core/array.fs]: remove ``len1`` altogether here?

**Comment by dsyme** [comment]
This looks really good.  I've left some comments about places where ``let len = ...`` can be removed.  

We should accept this into F# regardless of whether the JIT gets fixed to address https://github.com/Microsoft/visualfsharp/issues/1419 (because FSHarp.Core.dll often gets used with current versions of the JIT)

**Comment by jackmott** [comment]
Thanks for the review @dsyme, I'll check those cases and make adjustments.

**Comment by jackmott** [comment]
That appveyor error looks to be unrealted, I think.

**Comment by dsyme** [comment]
@jackmott I restarted that AppVeyor build just in case

**Comment by jackmott** [comment]
Yeah it looks like it is that AsyncChoice thing that comes up sometimes.

**Comment by liboz** [comment]
Are there any numbers for how much of a difference this change makes?

**Comment by jackmott** [comment]
Sure. Comparing a pretty average case of a single function  `map` with a simple mapping function `(fun x -> x + 1)` the perf difference is below.  As the mapping function gets more instructions, the % difference will tend towards zero of course.

A small difference but the code gets simpler and it affects many of the fundamental functions like subUnchecked, map, create, and init. The core DLL also gets 1,024 bytes smaller.


```ini

Host Process Environment Information:
BenchmarkDotNet=v0.9.8.0
OS=Microsoft Windows NT 6.2.9200.0
Processor=Intel(R) Core(TM) i7-4712HQ CPU 2.30GHz, ProcessorCount=8
Frequency=2240908 ticks, Resolution=446.2477 ns, Timer=TSC
CLR=MS.NET 4.0.30319.42000, Arch=64-bit RELEASE [RyuJIT]
GC=Concurrent Workstation
JitModules=clrjit-v4.6.1590.0

Type=SIMDBenchmark  Mode=Throughput  Platform=X64  
Jit=RyuJit  GarbageCollection=Concurrent Workstation  

```
 Method |  Length |            Median |         StdDev | Scaled | Gen 0 | Gen 1 |  Gen 2 | Bytes Allocated/Op |
------- |-------- |------------------ |--------------- |------- |------ |------ |------- |------------------- |
    **Old** |      **10** |        **17.5030 ns** |      **0.5275 ns** |   **1.00** |  **0.05** |     **-** |      **-** |              **26.88** |
    New |      10 |        14.1205 ns |      0.4858 ns |   0.81 |  0.05 |     - |      - |              26.11 |
    **Old** |   **10000** |    **10,212.8762 ns** |    **118.7990 ns** |   **1.00** | **27.45** |     **-** |      **-** |          **15,484.71** |
    New |   10000 |     8,963.2690 ns |    329.8907 ns |   0.88 | 29.07 |     - |      - |          16,395.45 |
    **Old** | **1000000** | **2,078,782.0718 ns** | **25,255.4172 ns** |   **1.00** |     **-** |     **-** | **858.90** |       **1,854,656.87** |
    New | 1000000 | 1,956,501.4913 ns | 15,930.2483 ns |   0.94 |     - |     - | 832.00 |       1,806,253.44 |

File: src/fsharp/FSharp.Core/array.fs [hunk 7]
```diff
@@ -57,7 +57,7 @@ namespace Microsoft.FSharp.Collections
         let create (count:int) (x:'T) =
             if count < 0 then invalidArg "count" (SR.GetString(SR.inputMustBeNonNegative))
             let array: 'T[] = Array.zeroCreateUnchecked count
-            for i = 0 to Operators.Checked.(-) count 1 do // use checked arithmetic here to satisfy FxCop
+            for i = 0 to Operators.Checked.(-) array.Length 1 do // use checked arithmetic here to satisfy FxCop
```
Comment (forki) [inline src/fsharp/FSharp.Core/array.fs]: urghs. do we really need to Operators.Checked.(-) here?

Comment (jackmott) [inline src/fsharp/FSharp.Core/array.fs]: One of the functions that has it, has a comment about it being for FXCOP (I don't know what that is)
I did test with and without it, and it made no difference on that function.  I could test all of them though to be sure.

Comment (forki) [inline src/fsharp/FSharp.Core/array.fs]: Fxcop is a tool that checks libs if they conform to the framework
guidelines. I assume that at some point this was an Unchecked call and then
it got changed because some policy check failed.

But if it makes no perf hit then we should just use plain old + and -
operators and get rid of such questions.

Am 15.08.2016 15:01 schrieb "Jack Mott" <notifications@github.com>:

Comment (jackmott) [inline src/fsharp/FSharp.Core/array.fs]: Just checked on another function, no performance impact there either.   I'd be happy to get rid of all the Checked bits for readability purposes in this PR, or I could do another one. @KevinRansom  or @dsyme  any input here?

Comment (forki) [inline src/fsharp/FSharp.Core/array.fs]: I think you should do that in separate pr

Am 15.08.2016 15:16 schrieb "Jack Mott" <notifications@github.com>:

**Comment by jackmott** [comment]
I'm all set here.
